subsys/bluetooth/host_extensions/conn_evt_cb.c

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/hci.h>
+#include <bluetooth/evt_cb.h>
+#include <helpers/nrfx_gppi.h>
+#include <bluetooth/hci_vs_sdc.h>
+
+#include <nrfx_egu.h>
+
+#if defined(CONFIG_BT_CONN_EVT_CB_EGU_INST_EGU0)
+#define EGU_INST_IDX 0
+#elif defined(CONFIG_BT_CONN_EVT_CB_EGU_INST_EGU1)
+#define EGU_INST_IDX 1
+#elif defined(CONFIG_BT_CONN_EVT_CB_EGU_INST_EGU2)
+#define EGU_INST_IDX 2
+#elif defined(CONFIG_BT_CONN_EVT_CB_EGU_INST_EGU3)
+#define EGU_INST_IDX 3
+#elif defined(CONFIG_BT_CONN_EVT_CB_EGU_INST_EGU4)
+#define EGU_INST_IDX 4
+#elif defined(CONFIG_BT_CONN_EVT_CB_EGU_INST_EGU5)
+#define EGU_INST_IDX 5
+#elif defined(CONFIG_BT_CONN_EVT_CB_EGU_INST_EGU020)
+#define EGU_INST_IDX 020
+#else
+#error "Unknown EGU instance"
+#endif
+
+static nrfx_egu_t egu_inst = NRFX_EGU_INSTANCE(EGU_INST_IDX);
+
+struct conn_evt_cb_cfg {
+	struct bt_conn_evt_cb cb;
+	struct bt_conn *conn;
+	void *user_data;
+	uint32_t prepare_time_us;
+
+	uint8_t ppi_channel;
+	uint16_t conn_interval_us;
+};
+
+static struct conn_evt_cb_cfg cfg[CONFIG_BT_MAX_CONN];
+static struct k_work_delayable work_queue_items[CONFIG_BT_MAX_CONN];
+
+static void work_handler(struct k_work *work)
```

Comment:
I'm not sure if I understand what the concern is here. Can you elaborate?
The `on_disconnected` callback cancels any outstanding work for the given connection

Reply:
I'm thinking that these functions can be called from separate threads, which requires attention. The cancel can fail and then we might call the NULL pointer, pass a dangling conn, or reschedule the work, etc..

I remember now since Zephyr 3.7 the disconnected callback is invoked from the system work queue, so these functions will indeed naturally synchronize. But this is not true before Zephyr 3.7, and I'm not sure it's sound design to rely on this anyway.

It may also be possible to find some reasoning about timings of connection radio events and disconnections that shows it's impossible for these functions to run at the same time, but that's very complex, and I actually doubt there is a bulletproof argument. It's much less work to assume the worst.

My opinion is that we should never rely on a simple cancel. The cancel is purely a optimization, and we must assume the work will run, possibly racing the function that calls cancel.

From an API standpoint, what guarantees do we give about when the last moment the callback can run is?